### PR TITLE
[feat] add a docker stack

### PR DESCRIPTION
**Pull request by nikophil**
Hi @kbond this PR is finally ready!

let's say you clone a fresh version of the repository, you'll have this output

```shell
$ make validate
[docker build...]
[docker up...]
[main composer install...]
[cs-fixer composer install...]
==> cs-fixer runs <==
[psalm composer install...]
==> psalm runs <==
[phpunit composer install...]
==> phpunit runs <==

$ make validate
==> cs-fixer runs <==
==> psalm runs <==
==> phpunit runs <==
```

---

If not cloning from a fresh version, for the first run, you must do something like this, to start from a fresh version:
```shell
$ make clear
```

---

I've added an entry `How to contribute` in the `README.md` file.It seems a better place than in symfony.com's doc

---

Thanks to @phpearth for this nice and understandable [documentation on Makefile + composer](https://docs.php.earth/interop/make/)

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        ports:
+            - "27017:27017"
+        volumes:
+            - ./docker/mongo-init.js:/docker-entrypoint-initdb.d/mongo-init.js:ro
+        healthcheck:
+            test: echo 'db.runCommand("ping").ok' | mongo mongo:27017/test --quiet
+            timeout: 10s
+            retries: 10
+
+volumes:
```
Comment (kbond) [inline docker-compose.yaml:71]: Forgive my docker ignorance but why is this required for postgres but not mysql?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+bin/tools/psalm/vendor/vimeo/psalm/psalm: vendor/autoload.php
+	${DOCKER_PHP} composer bin psalm install
+
+cs-fixer: bin/tools/cs-fixer/vendor/friendsofphp/php-cs-fixer/php-cs-fixer ### run PHP CS-Fixer
+	${DOCKER_PHP} bin/tools/cs-fixer/vendor/friendsofphp/php-cs-fixer/php-cs-fixer --no-interaction --diff -v fix --config=.php-cs-fixer.dist.php
+
+bin/tools/cs-fixer/vendor/friendsofphp/php-cs-fixer/php-cs-fixer: vendor/autoload.php
+	${DOCKER_PHP} composer bin cs-fixer install
+
+vendor/autoload.php:
```
Comment (kbond) [inline Makefile:60]: Possible to auto-run `docker-start` if not already running?

File: docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    mysql:
+        container_name: foundry_mysql
+        image: mysql
+        command: --default-authentication-plugin=mysql_native_password
+        environment:
+            MYSQL_ROOT_PASSWORD: root
+            MYSQL_DATABASE: zenstruck_foundry
+        ports:
+            - "3306:3306"
```
Comment (kbond) [inline docker-compose.yaml:28]: Should we use a non-standard local port? (same with postgres/mongo below)

Comment (nikophil) [reply docker-compose.yaml:28]: I'd like to provide the way to use a .env and to let the user chose. We can actually say `${MYSQL_PORT:-3306}`

File: Makefile (hunk 2 above)
Comment (nikophil) [reply Makefile:60]: yes but... I need to find a way to detect if it is already running. Makefile syntax is a bit cumbersome

File: bin/tools/cs-fixer/composer.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+    "require": {
+        "friendsofphp/php-cs-fixer": "^3.4"
```
[the remark this answers is not in the document]
Comment (nikophil) [reply bin/tools/cs-fixer/composer.json:3]: :+1:

File: docker-compose.yaml (hunk 1 above)
Comment (nikophil) [reply docker-compose.yaml:71]: this is not needed, it's useful to keep data when shutting down containers. I'll remove this, we don't need this

File: docker-compose.yaml (hunk 3 above)
Comment (kbond) [reply docker-compose.yaml:28]: Interesting, I'm not aware of that syntax, does that mean "use MYSQL_PORT if defined, otherwise use 3306"?

**Comment by nikophil** [comment]
@kbond BTW here is what I stil need to do:
- I'm struggling with psalm, I think we've introduced some SA errors with the sequence stuff.
- there is currently problem with mongo
- I'd like to provide a way to configure docker's port for users
- I need to find a way to call `docker-start` only when containers are not up

File: docker-compose.yaml (hunk 3 above)
Comment (nikophil) [reply docker-compose.yaml:28]: absolutely. This syntax comes from bash

**Comment by kbond** [comment]
> I'm struggling with psalm, I think we've introduced some SA errors with the sequence stuff.

The master branch runs fine. Is it because you upgraded psalm in this PR?

**Comment by nikophil** [comment]
> The master branch runs fine. Is it because you upgraded psalm in this PR?

right! it's OK now! but we'll have some troubles if we upgrade: return type of FactoryCollection's named constructor are not well php-documented (I think even phpstan will yell at us!)

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -376,4 +394,25 @@ jobs:
...
+
+  docker-stack:
+    name: CI with docker stack
+    continue-on-error: true
+    strategy:
+      matrix:
+        os:
+          - ubuntu-latest
+          - macos-latest
+          # - windows-latest
```
Comment (nikophil) [inline .github/workflows/ci.yml:407]: :man_shrugging: 

I have zero skills in windows, so I cannot make this one work

File: bin/tools/psalm/composer.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
     "require": {
-        "vimeo/psalm": "^4.10",
+        "vimeo/psalm": "4.10",
```
Comment (nikophil) [inline bin/tools/psalm/composer.json:3]: I've fixed psalm's version, since psalm analysis fails with bigger versions

File: phpunit-dama-doctrine.xml.dist [hunk 7]
```diff
@@ -13,7 +13,7 @@
         <env name="KERNEL_CLASS" value="Zenstruck\Foundry\Tests\Fixtures\Kernel" />
         <env name="USE_DAMA_DOCTRINE_TEST_BUNDLE" value="1" />
         <env name="SYMFONY_DEPRECATIONS_HELPER" value="max[self]=0&amp;max[direct]=0"/>
-        <env name="SHELL_VERBOSITY" value="-1"/>
+        <env name="SHELL_VERBOSITY" value="0"/>
```
Comment (nikophil) [inline phpunit-dama-doctrine.xml.dist:16]: I had hard time making the test suite work with mongo + php on docker because a SSL library was missing in php's Dockerfile.

Problem is with the error was not displaying anywhere. Here was the output of the error:
```shell
$ make phpunit-mongo
PHPUnit 8.5.30 #StandWithUkraine

Testing Project Test Suite
SSSSSSSSSSSSSSSSSSSSE

Time: 507 ms, Memory: 32.00 MB

There was 1 error:

1) Zenstruck\Foundry\Tests\Functional\FactoryDoctrineCascadeTest::many_to_one_relationship
RuntimeException: Error running "doctrine:mongodb:schema:create": 

/app/src/Test/AbstractSchemaResetter.php:24
/app/src/Test/ODMSchemaResetter.php:42
/app/src/Test/DatabaseResetter.php:53
/app/src/Test/ResetDatabase.php:60
```

This is because of `SHELL_VERBOSITY=-1`: `AbstractSchemaResetter` uses command's output.
With shell verbosity=0 you now can see the error:
```shell
$ make phpunit-mongo
PHPUnit 8.5.30 #StandWithUkraine

Testing Project Test Suite
SSSSSSSSSSSSSSSSSSSSE

Time: 507 ms, Memory: 32.00 MB

There was 1 error:

1) Zenstruck\Foundry\Tests\Functional\FactoryDoctrineCascadeTest::many_to_one_relationship
RuntimeException: Error running "doctrine:mongodb:schema:create": The SCRAM_SHA_256 authentication mechanism requires libmongoc built with ENABLE_SSL

/app/src/Test/AbstractSchemaResetter.php:24
/app/src/Test/ODMSchemaResetter.php:42
/app/src/Test/DatabaseResetter.php:53
/app/src/Test/ResetDatabase.php:60
```

File: docker-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        image: mysql
+        command: --default-authentication-plugin=mysql_native_password
+        environment:
+            MYSQL_ROOT_PASSWORD: root
+            MYSQL_DATABASE: zenstruck_foundry
+        ports:
+            - ${MYSQL_PORT:-3306}:3306
+        healthcheck:
+            test: 'mysql -u root -proot -D zenstruck_foundry -e "SELECT 1"'
+            timeout: 120s
```
Comment (nikophil) [inline docker-compose.yaml:31]: this was the only way I found to make ci with macos pass. Otherwise I had "mysql is not healthy" errors

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+ARGS := $(wordlist 2,$(words $(MAKECMDGOALS)),$(MAKECMDGOALS))
+$(eval $(RUN_ARGS):;@:)
+
+help:
+	@fgrep -h "###" $(MAKEFILE_LIST) | fgrep -v fgrep | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+
+tests: cs-fixer psalm phpunit-full ### Run full tests suite
+
+phpunit-full: vendor ### Run full PHPunit (MySQL + Mongo)
```
Comment (kbond) [inline Makefile:31]: I'm still questioning the need for all the permutations. I was sort of thinking just `test-full` (mongo/mysql in docker) and `test-fast` (sqlite no docker).

I don't feel super strongly about this but do you think it's important to have all these?

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	@$(eval filter ?= '.')
+	@${DC_EXEC} -e USE_FOUNDRY_BUNDLE=1 -e DATABASE_URL=${MYSQL_URL} -e MONGO_URL=${MONGO_URL} php vendor/bin/simple-phpunit --configuration phpunit-dama-doctrine.xml.dist --filter=$(filter)
+
+phpunit-fast: vendor ### Run PHPunit with SQLite
+	@$(eval filter ?= '.')
+	@${DC_EXEC} -e DATABASE_URL=${SQLITE_URL} php vendor/bin/simple-phpunit --configuration phpunit-dama-doctrine.xml.dist --filter=$(filter)
```
Comment (kbond) [inline Makefile:37]: I need to take a fresh look at https://github.com/zenstruck/foundry/pull/49. Would be nice to be able to run bundle tests with this too.

**Review by kbond — CHANGES_REQUESTED** [review]
Looking good!

Any thoughts on a `make migrations`? (I'm fine looking into this in another PR)

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,81 @@
+.PHONY: $(filter-out vendor bin/tools/psalm/vendor bin/tools/cs-fixer/vendor,$(MAKECMDGOALS))
+
+MYSQL_URL="mysql://root:root@mysql:3306/zenstruck_foundry?charset=utf8"
+POSTGRESQL_URL="postgresql://root:root@postgres:5432/zenstruck_foundry?charset=utf8"
+MONGO_URL="mongodb://mongo:mongo@mongo:27017/mongo?compressors=disabled&amp;gssapiServiceName=mongodb&authSource=mongo"
+SQLITE_URL="sqlite://./var/app.db"
+
+INTERACTIVE:=$(shell [ -t 0 ] && echo 1)
+ifdef INTERACTIVE
+	DC_EXEC=docker-compose exec
```
[the remark this answers is not in the document]
Comment (nikophil) [reply Makefile:10]: yes for sure. `docker compose` is actually `docker-compose v2` I think it will gradually replace it. 
I guess I can make something like `docker --help | grep compose` to know which one to use.

File: Makefile (hunk 9 above)
Comment (nikophil) [reply Makefile:31]: what if the CI fails on postgre? the user won't be able to reproduce...

File: phpunit-dama-doctrine.xml.dist (hunk 7 above)
Comment (nikophil) [reply phpunit-dama-doctrine.xml.dist:16]: yes :tada: 

the trick was to add some ssl libraries in the Dockerfile. But I recommend to keep the `SHELL_VERBOSITY=0`, since it does not pollute the tests with extra output, but we can see potential errors of doctrine commands under the hood

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+DOCKER_IS_UP = $(shell docker-compose ps | grep "Up (healthy)")
+
+ARGS := $(wordlist 2,$(words $(MAKECMDGOALS)),$(MAKECMDGOALS))
+$(eval $(RUN_ARGS):;@:)
+
+help:
+	@fgrep -h "###" $(MAKEFILE_LIST) | fgrep -v fgrep | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+
+tests: cs-fixer psalm phpunit-full ### Run full tests suite
```
[the remark this answers is not in the document]
Comment (nikophil) [reply Makefile:29]: `validate` sounds good

File: Makefile (hunk 9 above)
Comment (nikophil) [reply Makefile:31]: yes, why not... maybe I should update the "how to contribute" to explain this

File: Makefile (hunk 10 above)
Comment (kbond) [reply Makefile:37]: I don't think it's the same issue. #49 fails even on the latest version.

**Comment by nikophil** [comment]
@kbond could you tell me more about the `make migrations`? I'm not aware what is a need of migrations in foundry?

maybe we could open an issue for this, and do it in another PR?

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+phpunit-full: vendor ### Run full PHPunit (MySQL + Mongo)
+	@$(eval filter ?= '.')
+	@${DC_EXEC} -e USE_FOUNDRY_BUNDLE=1 -e DATABASE_URL=${MYSQL_URL} -e MONGO_URL=${MONGO_URL} php vendor/bin/simple-phpunit --configuration phpunit-dama-doctrine.xml.dist --filter=$(filter)
+
+phpunit-fast: vendor ### Run PHPunit with SQLite
```
Comment (kbond) [inline Makefile:35]: Can we switch the prefix from `phpunit-` to `test-` for these?

File: Makefile (hunk 10 above)
Comment (nikophil) [reply Makefile:37]: I'm closing this, since an issue exists.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -376,4 +394,25 @@ jobs:
...
+    steps:
+      - uses: docker-practice/actions-setup-docker@master
+        if: ${{ matrix.os == 'macos-latest' }}
+
+      - name: Checkout code
+        uses: actions/checkout@v2.3.3
+
+      - name: Run test suite with docker
+        run: |
+          make tests
```
Comment (kbond) [inline .github/workflows/ci.yml:418]:
```suggestion
          make validate
```

Check the gh actions, this fails but doesn't fail the job..?

File: Makefile [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+	@$(eval filter ?= '.')
+	@${DC_EXEC} -e DATABASE_URL=${MYSQL_URL} php vendor/bin/simple-phpunit --configuration phpunit-dama-doctrine.xml.dist --filter=$(filter)
+
+test-postgresql: docker-start vendor ### Run PHPunit with postgreSQL
+	@$(eval filter ?= '.')
+	@${DC_EXEC} -e DATABASE_URL=${POSTGRESQL_URL} php vendor/bin/simple-phpunit --configuration phpunit-dama-doctrine.xml.dist --filter=$(filter)
+
+test-mongo: docker-start vendor ### Run PHPunit with Mongo
+	@$(eval filter ?= '.')
+	@${DC_EXEC} -e MONGO_URL=${MONGO_URL} php vendor/bin/simple-phpunit --filter=$(filter)
```
Comment (kbond) [inline Makefile:59]:
```suggestion
	@${DC_EXEC} -e MONGO_URL=${MONGO_URL} php vendor/bin/simple-phpunit --configuration phpunit.xml.dist --filter=$(filter)
```

I had an issue with this target because I had my own `phpunit.xml` that phpunit picked up on and used.

File: README.md [hunk 16]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+at first before running the tests.
+
+### Run tests
+
+The library is shipped with a `Makefile` to run tests.
+Each target will build and start the docker stack and install composer only if needed.
+
+```shell
+$ make help
+validate                       Run full tests suite
```
Comment (kbond) [inline README.md:47]:
```suggestion
validate                       Run fixcs, sca and full test suite
```

File: README.md [hunk 17]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+test-full                      Run full PHPunit (MySQL + Mongo)
```
Comment (kbond) [inline README.md:48]:
```suggestion
test-full                      Run full PHPUnit (MySQL + Mongo)
```

File: README.md [hunk 18]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+test-fast                      Run PHPunit with SQLite
```
Comment (kbond) [inline README.md:49]:
```suggestion
test-fast                      Run PHPUnit with SQLite
```

File: README.md [hunk 19]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+test-mysql                     Run PHPunit with mysql
```
Comment (kbond) [inline README.md:50]:
```suggestion
test-mysql                     Run PHPUnit with mysql
```

File: README.md [hunk 20]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+test-postgresql                Run PHPunit with postgreSQL
```
Comment (kbond) [inline README.md:51]:
```suggestion
test-postgresql                Run PHPUnit with postgreSQL
```

File: README.md [hunk 21]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+test-mongo                     Run PHPunit with Mongo
```
Comment (kbond) [inline README.md:52]:
```suggestion
test-mongo                     Run PHPUnit with Mongo
```

File: README.md [hunk 22]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+sca                            Run Psalm
+docker-start                   Build and run containers
+docker-stop                    Stop containers
+docker-purge                   Purge containers
+composer                       Run composer command
+clear                          Start from a fresh install (needed if vendors have already been installed with another php version)
+
+```
+
+You can run each `phpunit-*` target with a special argument `filter`:
```
Comment (kbond) [inline README.md:63]:
```suggestion
You can run each `test-*` target with a special argument `filter`:
```

File: README.md [hunk 23]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+```shell
+$ make test-phunit filter=FactoryTest
```
Comment (kbond) [inline README.md:65]:
```suggestion
$ make test-mysql filter=FactoryTest
```

File: README.md [hunk 24]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+```
+
+which will use phpunit's `--filter` option.
```
Comment (kbond) [inline README.md:68]:
```suggestion
which will use PHPUnit's `--filter` option.
```

File: README.md [hunk 25]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+
+### Change docker's ports
+
+You can create a `.env` file to change the ports used by docker:
+```dotenv
```
Comment (kbond) [inline README.md:73]: This is a nice touch!

Curious, how does this work? I don't see anything in the Makefile that loads this.

File: README.md [hunk 26]
```diff
@@ -27,6 +27,60 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+MYSQL_PORT=3307
+POSTGRES_PORT=5434
+MONGO_PORT=27018
+```
+
+### Run tests without docker
+
+If for any reason docker is not available on your computer, the target `make test-fast` will run tests with your local
+php version, and sqlite will be used as SGDB. Results may differ from the CI!
```
Comment (kbond) [inline README.md:82]: SGDB?

**Review by kbond — CHANGES_REQUESTED** [review]
This works great!!! I'm so impressed!

Just a few minor things left.

File: .github/workflows/ci.yml (hunk 14 above)
Comment (nikophil) [reply .github/workflows/ci.yml:418]: yep I added `continue-on-error: true` to see the full result of the matrix. I've just removed it

File: README.md (hunk 25 above)
Comment (nikophil) [reply README.md:73]: docker-composer autoaticaly loads .env file if it exists

File: README.md (hunk 26 above)
Comment (nikophil) [reply README.md:82]: haha this is a french acronym! I was thinking it was an international one!

fixed

File: README.md [hunk 27]
```diff
@@ -27,6 +27,73 @@ Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/fou
...
+```shell
+$ docker-compose exec php [you commmand] # or "docker compose" depending on your docker compose version
+```
+
+### Using xdebug with PhpStorm
+
+The php container is shipped with xdebug activated. You can use step by step debugging session with PhpStorm: you should
+create a server called `FOUNDRY` in your PHP Remote Debug, with the IDE key `xdebug_foundry`
+
+![PhpStorm with xdebug](docs/phpstorm-xdebug-config.png)
```
Comment (nikophil) [inline README.md:90]: I don't know if it is useful :shrug:

**Review by kbond — APPROVED** [review]
